app/miragevpn_server_notun.ml

```diff
@@ -0,0 +1,255 @@
+  | None ->
+      Logs.err (fun m -> m "destination %a not found in map" Ipaddr.V4.pp dst);
+      Lwt.return_unit
+  | Some (fd, state) -> (
+      match Miragevpn.outgoing !state cs with
+      | Error `Not_ready ->
+          Logs.err (fun m ->
+              m "error not_ready while writing to %a" Ipaddr.V4.pp dst);
+          Lwt.return_unit
+      | Ok (state', enc) -> (
+          (* TODO fragmentation!? *)
+          state := state';
+          Common.write_to_fd fd enc >|= function
+          | Error (`Msg msg) ->
+              Logs.err (fun m ->
+                  m "%a tcp write failed %s" Ipaddr.V4.pp dst msg);
+              t.connections <- IPM.remove dst t.connections
+          | Ok () -> ()))
+
+let handle_payload t dst data =
+  match Ipv4_packet.Unmarshal.of_cstruct data with
+  | Error e ->
+      Logs.warn (fun m ->
+          m "%a received payload (error %s) %a" Ipaddr.V4.pp dst e
+            Cstruct.hexdump_pp data);
+      Lwt.return_unit
+  | Ok (ip, payload)
+    when ip.Ipv4_packet.proto = Ipv4_packet.Marshal.protocol_to_int `ICMP
+         && Ipaddr.V4.compare ip.Ipv4_packet.dst (fst t.ip) = 0 -> (
+      (* also check for icmp echo request! *)
+      match Icmpv4_packet.Unmarshal.of_cstruct payload with
+      | Ok (icmp, payload) ->
+          let reply = { icmp with Icmpv4_packet.ty = Icmpv4_wire.Echo_reply }
+          and ip' = { ip with src = ip.dst; dst = ip.src } in
+          let data =
+            Cstruct.append
+              (Icmpv4_packet.Marshal.make_cstruct ~payload reply)
+              payload
+          in
+          let hdr =
+            Ipv4_packet.Marshal.make_cstruct ~payload_len:(Cstruct.length data)
+              ip'
+          in
+          write t ip.src (Cstruct.append hdr data)
+      | Error e ->
+          Logs.warn (fun m -> m "ignoring icmp frame from, decoding error %s" e);
+          Lwt.return_unit)
+  | Ok (ip, _) ->
+      Logs.warn (fun m -> m "ignoring ipv4 frame %a" Ipv4_packet.pp ip);
+      Lwt.return_unit
+
+let rec timer server () =
+  let open Lwt.Infix in
+  (* foreach connection, call handle `Tick and follow instructions! *)
+  IPM.fold
+    (fun k (fd, t) acc ->
+      acc >>= fun acc ->
+      match Miragevpn.handle !t `Tick with
+      | Error e ->
+          Logs.err (fun m -> m "error in timer %a" Miragevpn.pp_error e);
+          Lwt.return acc
+      | Ok (_t', _out, _payloads, Some `Exit) ->
+          (* TODO anything to do with "_out" or "_payloads"? *)
+          Logs.warn (fun m -> m "exiting %a" Ipaddr.V4.pp k);
+          Lwt.return acc
+      | Ok (t', out, payloads, act) -> (
+          (* TODO anything to do with "_act"? (apart from exit) *)
+          Option.iter
+            (fun a ->
+              Logs.warn (fun m ->
+                  m "in timer, ignoring action %a" Miragevpn.pp_action a))
+            act;
+          t := t';
+          Lwt_list.iter_p (handle_payload server k) payloads >>= fun () ->
```

Comment:
I don't think we can produce payloads due to `` `Tick `` event. But it doesn't hurt (much) to iterate over the empty list.